js/store.js

```diff
@@ -296,58 +300,64 @@ const store = {
     }
     if (!thirdParty['firstPartyHostnames'].includes(origin)) {
       thirdParty['firstPartyHostnames'].push(origin);
-      isNewFirstParty = true;
     }
 
     // add link in first party if it doesn't exist yet
     // and the third party is visible (i.e. not whitelisted)
-    if (!('thirdPartyHostnames' in  firstParty)
-      || !firstParty['thirdPartyHostnames'].includes(target)) {
-      if (!('isVisible' in thirdParty) || !thirdParty['isVisible']) {
-        if (!(this.onWhiteList(origin, target))) {
-          // show third party
-          thirdParty['isVisible'] = true;
-          isNewThirdParty = true;
-        } else {
+    if (!this.isFirstPartyConnectedToThirdParty(firstParty, target)) {
+      if (!this.isVisibleThirdParty(thirdParty)) {
+        if (this.onWhiteList(origin, target)) {
           // hide third party
           thirdParty['isVisible'] = false;
+        } else {
+          // show third party and all its connections
+          thirdParty['isVisible'] = true;
+          isNewThirdParty = true;
+          thirdParty['firstPartyHostnames'].forEach((firstPartyHostname) => {
+            this.addFirstPartyLink(firstPartyHostname, target);
+          });
+          shouldUpdate = true;
         }
       }
-      if (thirdParty['isVisible'] && isNewFirstParty) {
+      if (this.isVisibleThirdParty(thirdParty) && !isNewThirdParty) {
         // an existing visible third party links to a new first party
-        thirdParty['firstPartyHostnames'].forEach((firstPartyHostname) => {
-          this.addFirstPartyLink(firstPartyHostname, target);
-        });
+        this.addFirstPartyLink(origin, target);
+        shouldUpdate = true;
```

Comment:
Again we might want to readress this later but yeah this logic seems correct now thanks.